### DISCUSSION TITLE
chapter_heap/top_k/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#552] [Announcements]
# chapter_heap/top_k/

一本动画图解、能运行、可提问的数据结构与算法入门书

https://www.hello-algo.com/chapter_heap/top_k/

**Comment by yuqiaoqiao** [comment]
第一次知道还有这么玩的，有点意思

**Comment by yuyu990116** [comment]
heapq.heappush(heap, nums[i])，这一步push之后不需要heapify吗

**Reply by krahets** [reply]
Hi 不需要哦，`heappush()` 入堆操作已经包含了堆化，`heapify()` 通常用于堆化输入列表的所有元素。

**Comment by zhuasdfg** [comment]
建长度为k的小顶堆实现top-K实在是泰裤辣

**Reply by marrblue** [reply]
优雅

**Comment by FlashSpiders** [comment]
```java
// 若当前元素大于堆顶元素，则将堆顶元素出堆、当前元素入堆
        if (nums[i] > heap.peek()) {
            heap.poll();
            heap.add(nums[i]);   // 求教，这里为什么是add不是用push()或者offer()呀，add的话也能加到堆顶吗
        }
```

**Reply by krahets** [reply]
Hi 在 Java 的 `PriorityQueue` 中，`add()` 实际上是在调用 `offer()` ，两者的表现是一致的。你会发现 Java 的常用数据结构基本没有用 `push()` 的，都是 `add()`

**Reply by lyl625760** [reply]
还是offer(e)更好一些， add更多和remove配合， offer更多和poll配合

**Comment by LyntNy4n** [comment]
建议解释一下为什么前k大的元素需要小根堆而不是大根堆 这里是个大坑

**Reply by krahets** [reply]
因为如果使用大根堆，我们每轮只能替换最大元素；而使用小根堆，我们每轮替换的是“第 K 大的元素”。

**Reply by xylithxxxx** [reply]
因为需要剔除最小的前n-k个数，最小堆顶部的元素就是最小的元素。当n个数中最小的n-k个数被从堆中删除后，剩下的就是最大的k个数，而最小堆顶部就是最大的k个数中最小的，也就是我们要的结果。

**Reply by primexiao** [reply]
估计想问的是，使用大顶堆，先入 n-k个元素，然后剩余的k个元素每次跟堆顶对比得到较大的一个，这样操作 k 次是不是也能得到 Top-k ？

**Comment by lyl625760** [comment]
太棒了

1. 利用上节内容，建堆代码可以优化下， 复杂度更低。

```java
// 将数组的前 k 个元素入堆
for (int i = 0; i < k; i++) {
    heap.offer(nums[i]);
}
```

可以优化成

```java
Queue<Integer> heap = new PriorityQueue<Integer>(Arrays.stream(nums).collect(Collectors.toList()));        
```

 2. 可以稍微提下 topk的其他解法，如快速排序

**Reply by krahets** [reply]
谢谢建议！

**Comment by palp1tate** [comment]
遍历和排序提供代码：
# 遍历
Python：

```python
def findKthLargest(nums, k):
    result = []
    for i in range(k):
        max_num = max(nums)
        result.append(max_num)
        nums.remove(max_num)
    return result
```

Go：

```go
func findKthLargest(nums []int, k int) []int {
	result := make([]int, k)
	for i := 0; i < k; i++ {
		max := nums[0]
		index := 0
		for i, num := range nums {
			if num > max {
				max = num
				index = i
			}
		}
		result[i] = max
		nums = append(nums[:index], nums[index+1:]...)
	}
	return result
}
```

# 排序
 Python：

```python
def findKthLargest(nums, k):
    nums.sort(reverse=True)
    return nums[:k]
```

Go：

```go
func findKthLargest(nums []int, k int) []int {
	sort.Sort(sort.Reverse(sort.IntSlice(nums)))
	return nums[:k]
}

//或者
func findKthLargest(nums []int, k int) []int {
	sort.Slice(nums, func(i, j int) bool {
		return nums[i] > nums[j]
	})
	return nums[:k]
}
```

**Comment by 5710xIa** [comment]
小顶堆是不是相当于会自动排序的数组，会把最小的数排到堆顶？不过如果K较大时，排数的过程会不会很耗时？

**Reply by krahets** [reply]
是的，由于 $k \leq n$ ，最差情况下时间复杂度为 $O(n \log n)$ 。这种情况实际上就是堆排序，在排序章节有讲

**Comment by T-1760** [comment]
想请教一下，为什么 JS / TS 与 其他编程语言处理方式不同呢？

JS / TS 中是采用 ： 先对 nums 列表j进行遍历为相反数，再通过 new MaxHeap() .从而实现初始化为 minHeap 小顶堆。 导致判断逻辑也与其他编程语言不同 `cur[i] < heap.peek()`

那 为什么不一开始就使用  new  MinHeap（），这样还节省了之前的列表遍历置反，后续又要置反输出最终结果，

**Reply by krahets** [reply]
Hi，因为这个 MaxHeap 是我们自己实现的。MinHeap 没有实现，因此需要用 MaxHeap 来模拟 MinHeap 的行为，具体做法就是将元素取反，这样做之后比较关系也完全镜像了

**Reply by T-1760** [reply]
嗯嗯，明白，谢谢

**Comment by CuiOrange** [comment]
请问代码实现部分的C语言为什么显示的是`[class]{}-[func]{topKHeap}`？是还没有上传代码吗？

**Reply by krahets** [reply]
是的，这部分 C 代码还待 PR 贡献

**Comment by Bbbbqsh** [comment]
大佬，想请教一下实现部分的python部分跟“把nums取负转成小顶堆，然后取k次堆顶元素，再取反”相比，哪个速度会更快呢？
```
def top_k_heap(nums: list[int], k: int) -> list[int]:
    l = []
    nums = [-i for i in nums]
    heapq.heapify(nums)
    for _ in range(k):
        l.append(heapq.heappop(nums) * -1)
    return l
```

**Reply by krahets** [reply]
先分析一下时间复杂度：

- 方法一：文中代码的时间复杂度是 $O(n \log k)$ 。
- 方法二：这段代码的时间复杂度是 $O(k \log n + n)$ ，其中复制列表 $O(n)$ ，建堆 $O(n)$ ，出堆 $O(k \log n)$ 。

最简单情况：当 $k = n$ 时，$n \log k < n \log k + n$ ，方法一稍快。
随着 $k$ 变小，从复杂度的角度估计就比较困难了，可能无法反映真实情况，实际性能要看具体的编程语言实现。

**Reply by Bbbbqsh** [reply]
明白了，感谢大佬 : )

**Reply by 480284856** [reply]
上述算法应该可以直接用大顶堆做吧，省去了复制列表的 $O(n)$ 时间：
使用全部数据直接初始化一个大顶堆，时间复杂度为 $O(n)$;
从上述大顶堆中选取 $k$ 个元素，时间复杂度为 $O(k\ log\ n) $

**Reply by sunshinesDL** [reply]
楼主的思路就是大顶堆

**Comment by wenjianmin** [comment]
是不是我理解有问题？如果第k+1个元素很大的话，是不是有问题？比如[100,1,2,3,4,5]，查找Top-3时，后面的元素都不会入堆

**Reply by wenjianmin** [reply]
描述有问题，应该说第一个堆顶元素很大时。

**Reply by Bbbbqsh** [reply]
heap[0]指的是堆顶元素，也就是[100, 1, 2]中的最小值1，而不是100

**Reply by wenjianmin** [reply]
是的，看了一下初始化的时候有进行堆化

**Comment by bnufw** [comment]
heapq 模块 中也提供了一样的nlargest方法

**Reply by bnufw** [reply]
我想问一下，这里的heap=[]，并没有将heap转化为堆，即heapq.heapify(heap)。虽然程序运行是没有错误的，但我还想知道为什么可以这样做？

**Reply by bnufw** [reply]
嗷嗷，堆就是用列表来表现的

**Comment by bnufw** [comment]
为什么是n轮出堆，不应该是n-k次么

**Reply by krahets** [reply]
你是对的～文中想表达的是“入堆+出堆”操作是在 n 轮循环中完成的，呈线性关系

**Comment by HangSeven** [comment]
注意： 元素入堆 这里应该指的是单个元素入堆的时间复杂度为O(log n），对于将n个元素入堆的总体平均时间复杂度前文也提到为O(nlog n）， 只不多通过建堆的操作优化可以将时间复杂度降低到O(n)
注意：元素入堆 这里指的是单个元素入堆的时间复杂度为O(log n), 对于遍历数组入堆的总体平均时间复杂度是O(n log n)， 只是通过遍历堆化的建堆操作的时间复杂度为0(n).

**Comment by acmu** [comment]
leetcode 原题：[215. 数组中的第K个最大元素](https://leetcode.cn/problems/kth-largest-element-in-an-array/description/?envType=list&envId=i5ohyv5)

**Reply by sunshinesDL** [reply]
还有道题，找最小的 k 个数： https://leetcode.cn/problems/smallest-k-lcci

**Comment by mochen-whu** [comment]
若当前元素大于堆顶元素，则将堆顶元素出堆、当前元素入堆
```python
if nums[i] > heap[0]:
      heapq.heappop(heap)
      heapq.heappush(heap, nums[i])
```
实际如果进一步追求性能，这里能否进行优化，将pop与push操作融合，即直接将当前元素替换堆顶元素，然后从堆顶向下堆化，这样可以省去一个堆化操作

**Reply by krahets** [reply]
确实，实际上有类似的操作 `heapq.heappushpop()` （不过执行顺序相反），效率会比分开执行高一些

**Reply by a2635814919** [reply]
@krahets 老师，我看heappushpop()方法是先入堆再弹出最小的，那是否可以初始化一个k+1大小的小顶堆，存入k个元素之和不用比较，直接执行heappushpop()方法，效率是否会高一点，还有更符合先出堆后入堆操作的heapreplace()需要进行判断。python学的不精不是很确定

**Comment by a2635814919** [comment]
想询问一下8.3 Top-k问题中方法三为什么C++的函数传入的是个引用，这边对原来的数组没有进行修改，是出于效率问题还是出于习惯传入了一个引用

**Reply by krahets** [reply]
这个要根据具体情况分析。如果函数不会修改输入数据，那么一般传入引用就好。

**Comment by CcYcC8** [comment]
想问下C#的8.3.3示例代码，是忘记大顶堆取反了吗？没有很理解小顶堆怎么构造的
```csharp
for (int i = 0; i < k; i++) {
        heap.Enqueue(nums[i], nums[i]);
    }
```

**Reply by hpstory** [reply]
Hi @CcYcC8 ，C#的优先队列默认就是小顶堆

**Comment by upseem** [comment]
总结：    
1、小顶堆找最大K个元素：维护一个大小为K的小顶堆，堆顶元素是当前K个元素中最小的。当遇到比堆顶元素大的新元素时，替换堆顶元素并重新调整堆。
    2、大顶堆找最小K个元素：维护一个大小为K的大顶堆，堆顶元素是当前K个元素中最大的。当遇到比堆顶元素小的新元素时，替换堆顶元素并重新调整堆。

**Comment by owenwu95** [comment]
C語言示例中的`int *getMinHeap`好像重複了

**Comment by tshzhu** [comment]
NOTE：用`heap.heapify`和`heap.heappushpop`应该能提高一点效率

**Comment by Asce-ops** [comment]
将数组的前 k 个元素入堆的时候可以考虑改用自下而上的堆化，时间复杂度为O(log k)；之后最多进行 n - k 轮出堆和 n - k 轮入堆，所以时间复杂度最多为O(log k + 2(n - k) · log k)

**Reply by sunshinesDL** [reply]
首先，前 k 个元素入堆，如果采取上一章的倒序遍历入堆法，是自下而上的建堆，而非自下而上的堆化；
其次，这种建堆方法的时间复杂度为O（n），因此前k个元素入堆的复杂度应为 O（k）。

**Reply by jiuerya7** [reply]
是自下而上对每个非叶节点进行堆化, 但是采用的堆化方式是自顶向下

**Comment by daichang01** [comment]
top-k完整实现#include <iostream>
```cpp
#include <vector>
#include <queue>



using namespace std;

vector<int> top_k(vector<int>& nums, int k) {
    priority_queue<int, vector<int>, greater<int>> min_Heap;
    for (int i = 0; i < k ;i++) {
        min_Heap.push(nums[i]);
    }
    for (int i = k; i < nums.size(); i++) {
        int top = min_Heap.top();
        if (nums[i] > top) {
            min_Heap.pop();
            min_Heap.push(nums[i]);
        }
    }
    vector<int> res;
    while (!min_Heap.empty()) {
        res.push_back(min_Heap.top());
        min_Heap.pop();
    }
    return res;
}

int main() {
    cout << "input n and k, requeire k < n:" << endl;
    int n , k;
    cin >> n >> k;
    int num;
    vector<int> nums(n);
    cout << "input n ints:" << endl;
    for (int i = 0; i < n; i++) {
        cin >> nums[i];
    }
    vector<int> res = top_k(nums, k);
    cout << "top-k:";
    for (int i : res) {
        cout << i << " ";
    }
    return 0;
}
```

**Comment by tstanon** [comment]
top_k最大k个元素的C语言版本代码
```c
void top_k(float *arr, int size, float *res, int k) {
    if (k <= 0 || k > size) return;
    float *heap = (float *)malloc(sizeof(float) * k); int heap_size = 0;
    for (int i = 0; i < k; i++){ pushHeap(heap, &heap_size, arr[i]); } //  // 初始化堆，放入前 k 个元素
    for (int i = k; i < size; i++){ if (arr[i] > heap[0]) { float tmp; popHeap(heap, &heap_size, &tmp); pushHeap(heap, &heap_size, arr[i]); } } // 处理剩余的元素
    for (int i = 0; i < k; i++) { popHeap(heap, &heap_size, &res[k - 1 - i]); } // 将堆中的元素按降序输出到 res 数组
    freeHeap(&heap);
}
```

**Comment by Ares-bit** [comment]
这C写的topk肯定有问题啊，怎么能初始化一个长度为0的heap然后往里push呢，这肯定段错误啊

**Reply by stazhenghaofu** [reply]
只加了3个啊，，全部比较完，最小的肯定在上面啊

**Reply by MuMingYi** [reply]
没找到`MaxHeap`这个结构体的具体定义（可能是我眼瞎），猜测应该是有个`MaxHeap::data[MAX_SIZE]`的成员，所以`newMaxHeap()`函数不论入参如何都会申请固定大小（`sizeof(MaxHeap)`）的内存用于存储节点

**Reply by Jacksonm75** [reply]
该章节源码my_heap.c中预分配好了data数组大小，长度为0只是初始没有往堆里拷贝任何数组数据，而是将前K个逐个插入堆中。
#define MAX_SIZE 5000
```c

/* 大顶堆 */
typedef struct {
    // size 代表的是实际元素的个数
    int size;
    // 使用预先分配内存的数组，避免扩容
    int data[MAX_SIZE];
} MaxHeap;
```

**Comment by YlguYtrid** [comment]
两个建议，一是先直接拷贝前k个元素再heapify，初始化速度更快；二是python的heap有PopPush操作，就是先Pop再Push，但是只需要logN时间，分别操作需要2logN时间，函数名是heapq.replace()。
```Python []
def top_k_heap(nums: list[int], k: int) -> list[int]:
    """基于堆查找数组中最大的 k 个元素"""
    # 初始化小顶堆
    heap = nums[:k]
    heapq.heapify(heap)
    # 从第 k+1 个元素开始，保持堆的长度为 k
    for i in range(k, len(nums)):
        # 若当前元素大于堆顶元素，则将堆顶元素出堆、当前元素入堆
        if nums[i] > heap[0]:
            heapq.replace(heap, nums[i])
    return heap
```

**Comment by YlguYtrid** [comment]
```Python []
def top_k_heap(nums: list[int], k: int) -> list[int]:
    """基于堆查找数组中最大的 k 个元素"""
    # 初始化小顶堆
    heap = nums[:k]
    heapq.heapify(heap)
    # 从第 k+1 个元素开始，保持堆的长度为 k
    for num in nums[k:]:
        # 若当前元素大于堆顶元素，则将堆顶元素出堆、当前元素入堆
        if nums > heap[0]:
            heapq.replace(heap, nums)
    return heap
```

**Comment by Zhaopudark** [comment]
这里，**遍历选择**[¶](https://www.hello-algo.com/chapter_heap/top_k/#831) 和 **堆**[¶](https://www.hello-algo.com/chapter_heap/top_k/#833) 方法，给出的时间复杂的关系、比较和解释，感觉不是很容易理解。因为很容易将 $k$  视为常数，那么推导出的时间复杂度就不一样了。

我的个人理解是：需要考虑的是一个二元函数 $T(n,k)$ 在约束 $k\le n$  条件下的渐近上界。

- 当使用**遍历选择**方法时：
    $T(n,k)=O(\frac{k[n+(n-k+1)]}{2}+k)$
    解释：
    
    - 假设列表元素中不存在相等值
    - 第1次遍历n个元素，提取1个最大的元素。
        - 使用一种 $O(1)$ 的方式让下次遍历时跳过已经提取的元素。
        - 比如使用链表实现这个数组，提取后直接删除
        - 或者使用哈希表记录已经提取的元素下标，下次遍历时删除
    - 第2次遍历剩下的n-1个元素，提取1个最大的元素。
    - …
    - 第k次遍历剩下的n-k+1个元素，提取1个最大的元素。
    - 总得操作次数，就是 $n+n-1+\cdots+n-k+1$ 这个等差数量的和，并补上这k次操作中，跳过已提取元素的 $O(1)$ 次操作，即常数项操作。
    
    所以有:
    $T(n,k)=O(\frac{k[n+(n-k+1)]}{2}+k)=O(n^2), \text{where } k\le n$
    
- 当使用**堆**方法时：
      $T(n,k)=O(n\log k)=O(n\log n) , \text{where } k\le n$

我的想法是，如果强调 $k$ 为常数，按照 “**函数渐近上界[¶](https://www.hello-algo.com/chapter_computational_complexity/time_complexity/#232)**” 中给出的一元函数渐近上界计算，那么上面两个时间复杂度都会变成 $O(n)$ 。那么很不好说，很不好体现堆方法的优势。那为了表达意思，就不得不使用 $<<$  等符号来进一步描述。至少在我的视角下，这是更难理解的。

不如咱么就计算一个二元函数的渐近上界。那么，堆方法的时间复杂度 $O(n\log n)$ 就比遍历选择方法的 $O(n^2)$ 更小。这个优势就非常好说。

进一步解释的话：

- Top-k 问题的算法操作次数同时和 $n$ 与 $k$ 有关。操作次数 $T(n,k)$ 是一个二元函数，满足 $k\le n$ 的约束。
- 考虑时间复杂度时，应当考虑 $T(n,k)$ 的渐近上界 $f(n,k)$
- 如果没有 $k\le n$ 的约束，**遍历选择** 和 **堆** 方法 的时间复杂度 就将分别为 $O(nk)$ 和 $O(n\log k)$。而有了 $k\le n$  约束后，上述时间复杂度就可以退化为 $O(n^2)$ 和 $O(n\log n)$
- 对于 **排序** 方法，无论有没有 $k\le n$ 的约束，时间复杂度都与 $k$ 无关，都是 $O(n\log n)$

**Comment by xiaozhuge123** [comment]
day07

**Comment by linga2006** [comment]
我想请教大家，如果用堆的方法做top-k问题的话，由于堆内元素是无序的，所以最后输出的序列定不准是啥，但编程题的答案肯定是个确定序列，这该咋办呢    那如果是最后再排一次序的话，那和先排序再取k个元素又有啥区别呢，又怎么能体现堆做法效率最高呢，请大家帮忙解惑，谢谢

**Reply by linga2006** [reply]
好吧我知道了，k<n，效率依然最高
